File: src/lexer/lexer.cpp
#include "lexer.hpp"

void enumerate(const TokenType &type) {
  switch (type) {
    case TokenType::ID:
      std::cout << "ID";
      break;
    case TokenType::PUB:
      std::cout << "PUB";
      break;
    case TokenType::FN:
      std::cout << "FN";
      break;
    case TokenType::STRUCT:
      std::cout << "STRUCT";
      break;
    case TokenType::ENUM:
      std::cout << "ENUM";
      break;
    case TokenType::IMPL:
      std::cout << "IMPL";
      break;
    case TokenType::IF:
      std::cout << "IF";
      break;
    case TokenType::ELSE:
      std::cout << "ELSE";
      break;
    case TokenType::FOR:
      std::cout << "FOR";
      break;
    case TokenType::WHILE:
      std::cout << "WHILE";
      break;
    case TokenType::INTEGER_LIT:
      std::cout << "INTEGER_LIT";
      break;
    case TokenType::PLUS:
      std::cout << "PLUS";
      break;
    case TokenType::MINUS:
      std::cout << "MINUS";
      break;
    case TokenType::ASTERISK:
      std::cout << "ASTERISK";
      break;
    case TokenType::FSLASH:
      std::cout << "FSLASH";
      break;
    case TokenType::SEMICOLON:
      std::cout << "SEMICOLON";
      break;
    case TokenType::LPAR:
      std::cout << "LPAR";
      break;
    case TokenType::RPAR:
      std::cout << "RPAR";
      break;
    case TokenType::LCURL:
      std::cout << "LCURL";
      break;
    case TokenType::RCURL:
      std::cout << "RCURL";
      break;
    case TokenType::ARROW:
      std::cout << "ARROW";
      break;
    case TokenType::CHAR:
      std::cout << "CHAR";
      break;
    case TokenType::I32:
      std::cout << "I32";
      break;
    case TokenType::RET:
      std::cout << "RET";
      break;
    case TokenType::COLON:
      std::cout << "COLON";
      break;
    case TokenType::UNKNOWN:
    default:
      std::cout << "UNKNOWN";
      break;
  }
}

bool is_num(const TokenType &type) {
  switch (type) {
    case TokenType::INTEGER_LIT:
      return true;
    default:
      return false;
  }
}

bool is_operator(const TokenType &type) {
  switch (type) {
    case TokenType::PLUS:
      return true;
    case TokenType::MINUS:
      return true;
    case TokenType::ASTERISK:
      return true;
    case TokenType::FSLASH:
      return true;
    default:
      return false;
  }
}

TokenType match(const std::string &lexeme) {
  if (lexeme == "fn") {
    return TokenType::FN;
  } else if (lexeme == "pub") {
    return TokenType::PUB;
  } else if (lexeme == "struct") {
    return TokenType::STRUCT;
  } else if (lexeme == "enum") {
    return TokenType::ENUM;
  } else if (lexeme == "impl") {
    return TokenType::IMPL;
  } else if (lexeme == "if") {
    return TokenType::IF;
  } else if (lexeme == "else") {
    return TokenType::ELSE;
  } else if (lexeme == "for") {
    return TokenType::FOR;
  } else if (lexeme == "char") {
    return TokenType::CHAR;
  } else if (lexeme == "i32") {
    return TokenType::I32;
  } else if (lexeme == "ret" || lexeme == "return") {
    return TokenType::RET;
  } else {
    return TokenType::ID;
  }
}

void print_tokens(const std::vector<Token> &_tokens) {
  for (const Token &token : _tokens) {
    std::cout << "[ ";
    enumerate(token.type);
    std::cout << " , \'" << token.lexeme << "\' ]\n";
  }
}

Lexer::Lexer(const std::string &content) {
  for (size_t i = 0; i < content.length(); ++i) {
    if (isspace(content[i]) || isblank(content[i])) {
      continue;
    } else if (isalpha(content[i]) || content[i] == '_') {
      std::string value;
      while (isalnum(content[i])) {
        value += static_cast<char>(content[i]);
        ++i;
      }
      --i;

      _tokens.emplace_back(Token{match(value), value});
    } else if (isdigit(content[i])) {
      std::string value;
      while (isdigit(content[i])) {
        value += static_cast<char>(content[i]);
        ++i;
      }
      --i;

      _tokens.emplace_back(Token {TokenType::INTEGER_LIT, value});
    } else if (!isalnum(content[i]) && !isdigit(content[i]) && isascii(content[i])) {
      std::string value(1, content[i]);
      if (content[i] == ';') {
        _tokens.emplace_back(Token{TokenType::SEMICOLON, value});
      } else if (content[i] == ':') {
        _tokens.emplace_back(Token{TokenType::COLON, value});
      } else if (content[i] == '*') {
        _tokens.emplace_back(Token{TokenType::ASTERISK, value});
      } else if (content[i] == '-') {
        if (content[i + 1] == '>') {
          _tokens.emplace_back(Token{TokenType::ARROW, "->"});
          i++;
        } else {
          _tokens.emplace_back(Token{TokenType::MINUS, value});
        }
      } else if (content[i] == '+') {
        _tokens.emplace_back(Token{TokenType::PLUS, value});
      } else if (content[i] == '(') {
        _tokens.emplace_back(Token{TokenType::LPAR, value});
      } else if (content[i] == ')') {
        _tokens.emplace_back(Token{TokenType::RPAR, value});
      } else if (content[i] == '{') {
        _tokens.emplace_back(Token{TokenType::LCURL, value});
      } else if (content[i] == '}') {
        _tokens.emplace_back(Token{TokenType::RCURL, value});
      } else if (content[i] == '/') {
        if (content[i + 1] == '*') {
          i += 2;
          while (content[i] != '*' && content[i + 1] != '/') i++;
            i++;
        } else if (content[i + 1] == '/') {
          i += 2;
          while (content[i] != '\n' && content[i] != '\r') i++;
        } else {
          _tokens.emplace_back(Token{TokenType::FSLASH, value});
        }
        
      } else {
        _tokens.emplace_back(Token{TokenType::UNKNOWN, value});
      }
    } else {
      _tokens.emplace_back(Token {TokenType::UNKNOWN, std::string(1, static_cast<char>(content[i]))});
    }
  }
}

Lexer::~Lexer() {}

const std::vector<Token>& Lexer::tokens() {
  return _tokens;
}

File: src/front/lexer.rs
use std::iter::Peekable;
use std::str::Chars;

use crate::front::token::Token;
use super::token::Position;

macro_rules! here {
    () => {
        println!(
            "Execution passed through here:\n\tfile: {}\n\tline: {}",
            file!(),
            line!()
        )
    };
}

pub struct Lexer<'a> {
    position: Position,
    input: Peekable<Chars<'a>>,
}

impl<'a> Iterator for Lexer<'a> {
    type Item = (Token, Position);

    fn next(&mut self) -> Option<Self::Item> {
        self.next_token_internal()
    }
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Lexer {
            position: Position { line: 1, index: 1 },
            input: input.chars().peekable(),
        }
    }

    fn update_position(&mut self, ch: char) {
        match ch {
            '\n' => {
                self.position.line += 1;
                self.position.index = 1; // Reset index on a new line
            }
            _ => {
                self.position.index += 1;
            }
        }
    }

    /// Skip a nested multiline comment.
    /// Returns true if the comment was terminated correctly, false otherwise.
    fn skip_multiline_comment(&mut self) -> bool {
        let mut depth = 1;
        while let Some(ch) = self.input.next() {
            self.update_position(ch);
            match ch {
                '/' => {
                    if let Some(&next_ch) = self.input.peek() {
                        if next_ch == '*' {
                            self.input.next(); // Consume '*'
                            self.update_position('*');
                            depth += 1;
                        }
                    }
                }
                '*' => {
                    if let Some(&next_ch) = self.input.peek() {
                        if next_ch == '/' {
                            self.input.next(); // Consume '/'
                            self.update_position('/');
                            depth -= 1;
                            if depth == 0 {
                                return true;
                            }
                        }
                    }
                }
                _ => {} // Continue consuming comment characters.
            }
        }
        // If we exit the loop, EOF was reached before closing all nested comments.
        false
    }

    fn next_token_internal(&mut self) -> Option<(Token, Position)> {
        while let Some(&ch) = self.input.peek() {
            // Skip whitespace
            match ch {
                ' ' | '\t' | '\r' => {
                    self.input.next();
                    self.update_position(ch);
                    continue;
                }
                '\n' => {
                    self.input.next();
                    self.update_position(ch);
                    continue;
                }
                // String literal start
                '\"' => {
                    return Some((self.string_literal(), self.position.clone()));
                }
                // Character literal start
                '\'' => {
                    return Some((self.character_literal(), self.position.clone()));
                }
                '0'..='9' => {
                    let pos = self.position.clone();
                    return Some((self.number(), pos));
                }
                'a'..='z' | 'A'..='Z' | '_' => {
                    let pos = self.position.clone();
                    return Some((self.identifier_or_keyword(), pos));
                }
                '+' => {
                    self.input.next();
                    self.update_position(ch);
                    return Some((Token::Plus, self.position.clone()));
                }
                '-' => {
                    self.input.next(); // Consume '-'
                    let pos = self.position.clone();
                    self.update_position(ch);
                    if let Some(&next_ch) = self.input.peek() {
                        if next_ch == '>' {
                            self.input.next();
                            self.update_position(next_ch);
                            return Some((Token::Arrow, pos));
                        }
                    }
                    return Some((Token::Minus, pos));
                }
                '&' => {
                    self.input.next(); // Consume '-'
                    let pos = self.position.clone();
                    self.update_position(ch);
                    if let Some(&next_ch) = self.input.peek() {
                        if next_ch == '&' {
                            self.input.next();
                            self.update_position(next_ch);
                            return Some((Token::And, pos));
                        }
                    }
                    return Some((Token::Ampersand, pos));
                }
                '|' => {
                    self.input.next(); // Consume '-'
                    let pos = self.position.clone();
                    self.update_position(ch);
                    if let Some(&next_ch) = self.input.peek() {
                        if next_ch == '&' {
                            self.input.next();
                            self.update_position(next_ch);
                            return Some((Token::Or, pos));
                        }
                    }
                    return Some((Token::Pipe, pos));
                }
                '*' => {
                    self.input.next();
                    self.update_position(ch);
                    return Some((Token::Asterisk, self.position.clone()));
                }
                '/' => {
                    self.input.next(); // Consume '/'
                    self.update_position(ch);
                    if let Some(&next_ch) = self.input.peek() {
                        if next_ch == '/' {
                            self.input.next(); // Consume the second '/'
                            self.update_position(next_ch);
                            // Skip single-line comment
                            while let Some(&comment_ch) = self.input.peek() {
                                if comment_ch == '\n' {
                                    break;
                                }
                                self.input.next();
                                self.update_position(comment_ch);
                            }
                            continue; // Restart the loop after comment.
                        } else if next_ch == '*' {
                            self.input.next(); // Consume '*' signaling multiline comment.
                            self.update_position(next_ch);
                            // Skip the entire multiline comment.
                            if !self.skip_multiline_comment() {
                                println!("Error: Unterminated multiline comment.");
                                // In a real compiler, you might return an Error token or panic.
                            }
                            continue; // Restart scanning tokens after the comment.
                        }
                    }
                    return Some((Token::Fslash, self.position.clone()));
                }
                '%' => {
                    self.input.next();
                    self.update_position(ch);
                    return Some((Token::Percent, self.position.clone()));
                }
                '=' => {
                    self.input.next();
                    self.update_position(ch);
                    return Some((Token::Equal, self.position.clone()));
                }
                '(' => {
                    self.input.next();
                    self.update_position(ch);
                    return Some((Token::LPar, self.position.clone()));
                }
                ')' => {
                    self.input.next();
                    self.update_position(ch);
                    return Some((Token::RPar, self.position.clone()));
                }
                '{' => {
                    self.input.next();
                    self.update_position(ch);
                    return Some((Token::LCurl, self.position.clone()));
                }
                '}' => {
                    self.input.next();
                    self.update_position(ch);
                    return Some((Token::RCurl, self.position.clone()));
                }
                ',' => {
                    self.input.next();
                    self.update_position(ch);
                    return Some((Token::Comma, self.position.clone()));
                }
                ';' => {
                    self.input.next();
                    self.update_position(ch);
                    return Some((Token::Semicolon, self.position.clone()));
                }
                ':' => {
                    self.input.next();
                    let pos = self.position.clone();
                    self.update_position(ch);
                    if let Some(&next_ch) = self.input.peek() {
                        if next_ch == '=' {
                            self.input.next(); // Consume '='
                            self.update_position(next_ch);
                            return Some((Token::Walrus, pos));
                        }
                    }
                    return Some((Token::Colon, pos));
                }
                _ => {
                    self.input.next();
                    self.update_position(ch);
                    println!("Warning: Unknown token '{}'", ch);
                    return Some((Token::Unknown(ch), self.position.clone()));
                }
            }
        }
        None
    }

    fn number(&mut self) -> Token {
        let mut num_str = String::new();
        let mut has_decimal = false;

        while let Some(&ch) = self.input.peek() {
            if ch.is_ascii_digit() {
                num_str.push(ch);
                self.input.next(); // Consume digit
                self.update_position(ch);
            } else if ch == '.' {
                if has_decimal || self.input.clone().nth(1) == Some('.') {
                    break;
                }
                has_decimal = true;
                num_str.push(ch);
                self.input.next(); // Consume '.'
                self.update_position(ch);
            } else {
                break;
            }
        }

        Token::NumberLiteral(num_str)
    }

    fn identifier_or_keyword(&mut self) -> Token {
        let mut ident = String::new();

        while let Some(&ch) = self.input.peek() {
            if ch.is_ascii_alphanumeric() || ch == '_' {
                ident.push(ch);
                self.input.next(); // Consume character
                self.update_position(ch);
            } else {
                break;
            }
        }

        match ident.as_str() {
            "fn" => Token::Fn,
            "ret" => Token::Ret,
            "struct" => Token::Struct,
            "pub" => Token::Pub,
            "enum" => Token::Enum,
            "impl" => Token::Impl,
            "if" => Token::If,
            "else" => Token::Else,
            "for" => Token::For,
            "while" => Token::While,
            "i32" => Token::I32,
            "i64" => Token::I64,
            "u32" => Token::U32,
            "u64" => Token::U64,
            "usize" => Token::Usize,
            "f32" => Token::F32,
            "f64" => Token::F64,
            "char" => Token::Char,
            _ => Token::Identifier(ident),
        }
    }

    fn string_literal(&mut self) -> Token {
        let mut literal = String::new();
        // Consume the opening double-quote.
        self.input.next();
        self.update_position('\"');
        
        while let Some(&ch) = self.input.peek() {
            if ch == '"' {
                self.input.next(); // Consume the closing quote.
                self.update_position(ch);
                break;
            } else if ch == '\\' {
                self.input.next();
                self.update_position('\\');
                if let Some(&esc_ch) = self.input.peek() {
                    let escaped_char = match esc_ch {
                        'n' => '\n',
                        't' => '\t',
                        'r' => '\r',
                        '"' => '"',
                        '\\' => '\\',
                        '\'' => '\'',
                        other => other,
                    };
                    literal.push(escaped_char);
                    self.input.next();
                    self.update_position(esc_ch);
                }
            } else {
                literal.push(ch);
                self.input.next();
                self.update_position(ch);
            }
        }
        Token::StringLiteral(literal)
    }

    fn character_literal(&mut self) -> Token {
        // Consume the opening single-quote.
        self.input.next();
        self.update_position('\'');
        let mut char_val = None;
        if let Some(&ch) = self.input.peek() {
            if ch == '\\' {
                self.input.next();
                self.update_position('\\');
                if let Some(&esc_ch) = self.input.peek() {
                    let c = match esc_ch {
                        'n' => '\n',
                        't' => '\t',
                        'r' => '\r',
                        '\\' => '\\',
                        '\'' => '\'',
                        '"' => '"',
                        other => other,
                    };
                    char_val = Some(c);
                    self.input.next();
                    self.update_position(esc_ch);
                }
            } else {
                char_val = Some(ch);
                self.input.next();
                self.update_position(ch);
            }
        }
        // Expect the closing single-quote.
        if let Some(&ch) = self.input.peek() {
            if ch == '\'' {
                self.input.next();
                self.update_position('\'');
            }
        }
        Token::CharacterLiteral(char_val.unwrap_or('\0').to_string())
    }

    pub fn lex(self) -> Vec<(Token, Position)> {
        let mut vec: Vec<(Token, Position)> = self.collect();
        vec.push((Token::Eof, Position { line: vec.last().unwrap().1.line + 1, index: 1 }));
        vec
    }
}
